File: bindings/php/src/ProvenSdn.php
<?php
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// SDN protocol types for proven-servers.

declare(strict_types=1);

namespace ProvenServers;

/** SdnMessageType matching the Idris2 ABI tags. */
enum SdnMessageType: int
{
    case Hello = 0;
    case Error = 1;
    case EchoRequest = 2;
    case EchoReply = 3;
    case FeaturesRequest = 4;
    case FeaturesReply = 5;
    case FlowMod = 6;
    case PacketIn = 7;
    case PacketOut = 8;
    case PortStatus = 9;
    case BarrierRequest = 10;
    case BarrierReply = 11;
}

/** FlowAction matching the Idris2 ABI tags. */
enum FlowAction: int
{
    case Output = 0;
    case SetField = 1;
    case Drop = 2;
    case PushVlan = 3;
    case PopVlan = 4;
    case SetQueue = 5;
    case Group = 6;
}

/** MatchField matching the Idris2 ABI tags. */
enum MatchField: int
{
    case InPort = 0;
    case EthDst = 1;
    case EthSrc = 2;
    case EthType = 3;
    case VlanId = 4;
    case IpSrc = 5;
    case IpDst = 6;
    case TcpSrc = 7;
    case TcpDst = 8;
    case UdpSrc = 9;
    case UdpDst = 10;
}

/** PortState matching the Idris2 ABI tags. */
enum PortState: int
{
    case Up = 0;
    case Down = 1;
    case Blocked = 2;
}

File: bindings/csharp/src/ProvenSdn.cs
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// SDN protocol types for proven-servers.

namespace Proven;

/// <summary>SdnMessageType matching the Idris2 ABI tags (0-11).</summary>
public enum SdnMessageType : byte
{
    Hello = 0,
    Error = 1,
    EchoRequest = 2,
    EchoReply = 3,
    FeaturesRequest = 4,
    FeaturesReply = 5,
    FlowMod = 6,
    PacketIn = 7,
    PacketOut = 8,
    PortStatus = 9,
    BarrierRequest = 10,
    BarrierReply = 11
}

/// <summary>FlowAction matching the Idris2 ABI tags (0-6).</summary>
public enum FlowAction : byte
{
    Output = 0,
    SetField = 1,
    Drop = 2,
    PushVlan = 3,
    PopVlan = 4,
    SetQueue = 5,
    Group = 6
}

/// <summary>MatchField matching the Idris2 ABI tags (0-10).</summary>
public enum MatchField : byte
{
    InPort = 0,
    EthDst = 1,
    EthSrc = 2,
    EthType = 3,
    VlanId = 4,
    IpSrc = 5,
    IpDst = 6,
    TcpSrc = 7,
    TcpDst = 8,
    UdpSrc = 9,
    UdpDst = 10
}

/// <summary>PortState matching the Idris2 ABI tags (0-2).</summary>
public enum PortState : byte
{
    Up = 0,
    Down = 1,
    Blocked = 2
}
